src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/x64/windows/WinVaList.java

```diff
@@ -0,0 +1,239 @@
+        segment.close();
+        copies.forEach(MemorySegment::close);
+    }
+
+    @Override
+    public CSupport.VaList copy() {
+        return WinVaList.ofAddress(ptr);
+    }
+
+    @Override
+    public MemoryAddress toAddress() {
+        return ptr;
+    }
+
+    @Override
+    public boolean isAlive() {
+        return segment.isAlive();
+    }
+
+    static class Builder implements CSupport.VaList.Builder {
+
+        private final List<SimpleVaArg> args = new ArrayList<>();
+
+        private Builder arg(Class<?> carrier, MemoryLayout layout, Object value) {
+            SharedUtils.checkCompatibleType(carrier, layout, Windowsx64Linker.ADDRESS_SIZE);
+            args.add(new SimpleVaArg(carrier, layout, value));
+            return this;
+        }
+
+        @Override
+        public Builder intArg(MemoryLayout layout, int value) {
+            return arg(int.class, layout, value);
+        }
+
+        @Override
+        public Builder longArg(MemoryLayout layout, long value) {
+            return arg(long.class, layout, value);
+        }
+
+        @Override
+        public Builder doubleArg(MemoryLayout layout, double value) {
+            return arg(double.class, layout, value);
+        }
+
+        @Override
+        public Builder memoryAddressArg(MemoryLayout layout, MemoryAddress value) {
+            return arg(MemoryAddress.class, layout, value);
+        }
+
+        @Override
+        public Builder memorySegmentArg(MemoryLayout layout, MemorySegment value) {
+            return arg(MemorySegment.class, layout, value);
+        }
+
+        public WinVaList build() {
+            MemorySegment ms = MemorySegment.allocateNative(VA_SLOT_SIZE_BYTES * args.size());
+            List<MemorySegment> copies = new ArrayList<>();
+
+            MemoryAddress addr = ms.baseAddress();
+            for (SimpleVaArg arg : args) {
+                if (arg.carrier == MemorySegment.class) {
+                    MemorySegment msArg = ((MemorySegment) arg.value);
+                    TypeClass typeClass = TypeClass.typeClassFor(arg.layout);
+                    switch (typeClass) {
+                        case STRUCT_REFERENCE -> {
+                            MemorySegment copy = MemorySegment.allocateNative(arg.layout);
+                            copy.copyFrom(msArg); // by-value
+                            copies.add(copy);
+                            VH_address.set(addr, copy.baseAddress());
+                        }
+                        case STRUCT_REGISTER -> {
+                            MemorySegment slice = ms.asSlice(addr.segmentOffset(), VA_SLOT_SIZE_BYTES);
+                            slice.copyFrom(msArg);
+                        }
+                        default -> throw new IllegalStateException("Unexpected TypeClass: " + typeClass);
+                    }
+                } else {
+                    VarHandle writer = arg.varHandle();
+                    writer.set(addr, arg.value);
```

Comment:
From a performance perspective, both this and the reader code above are suboptimal, in two dimensions:
* the VH used is not constant
* the type of the get/set call is not exact
No need to dive in premature optimization territory now, but perhaps worth keeping in mind.

Reply:
Thanks. I want to get the API right first, and then move on to tackling performance after that is finished, since diagnosing and solving performance problems usually takes a long time on it's own.